### PR TITLE
Add temporaryCacheLocation to Cache Options

### PR DESCRIPTION
**Pull request by nicolaszawada**
This is a PR follow up on [the issue we opened a few years ago](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4231).

So in summary, we added a new property to the `CacheOptions` to make the temporary cache storage location configurable. It defaults to the existing cache storage location so it doesn't break anything when not configured.

### CODE REVIEW

**Comment by nicolaszawada** [comment]
@microsoft-github-policy-service agree

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 1]
```diff
@@ -71,16 +70,30 @@ export class BrowserCacheManager extends CacheManager {
...
+            }
+        }
+    
+        switch (tempCacheLocation) {
             case BrowserCacheLocation.LocalStorage:
             case BrowserCacheLocation.SessionStorage:
                 try {
-                    // Temporary cache items will always be stored in session storage to mitigate problems caused by multiple tabs
-                    return new BrowserStorage(BrowserCacheLocation.SessionStorage);
+                    return new BrowserStorage(tempCacheLocation);
```
Comment (tnorling) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts]: I think all of this can be simplified with just the following change: 

```suggestion
                    return new BrowserStorage(temporaryCacheLocation || BrowserCacheLocation.SessionStorage);
```

File: lib/msal-browser/src/config/Configuration.ts [hunk 2]
```diff
@@ -77,6 +77,10 @@ export type CacheOptions = {
      * Used to specify the cacheLocation user wants to set. Valid values are "localStorage" and "sessionStorage"
      */
     cacheLocation?: BrowserCacheLocation | string;
+    /**
+     * Used to specify the temporaryCacheLocation user wants to set. Valid values are "localStorage" and "sessionStorage"
```
Comment (tnorling) [inline lib/msal-browser/src/config/Configuration.ts]: memoryStorage?

File: lib/msal-browser/docs/caching.md [hunk 3]
```diff
@@ -58,7 +58,7 @@ To faciliate efficient token acquisition while maintaining a good UX, MSAL cache
     - previous failed request 
     - performance data
 
-> :bulb: Temporary cache entries will always be stored in session storage or in memory. MSAL will fallback to memory storage if local/session storage is not available.
+> :bulb: Temporary cache entries will always be stored in session storage or in memory unless overridden by the user with the `temporaryCacheLocation` cache option. MSAL will fallback to memory storage if local/session storage is not available.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]: I think we will also want to clearly document the potential problems with overriding the temporary cache location. Specifically when choosing localStorage interaction in more than one tab/window will not be supported and you may receive interaction_in_progress errors unexpectedly. 

Please describe the scenario where this should be overridden. This is an escape hatch, not a fully supported feature.

File: change/@azure-msal-browser-25232dc8-c6d0-4238-ad59-1c1f3fcd1a7a.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add temporaryCacheLocation to Cache Options",
```
Comment (tnorling) [inline change/@azure-msal-browser-25232dc8-c6d0-4238-ad59-1c1f3fcd1a7a.json]:
```suggestion
  "comment": "Add temporaryCacheLocation to Cache Options #5725",
```

**Review by tnorling — COMMENTED** [review]
Thanks for the PR! Left a couple suggestions, would be great if you could include a couple tests as well.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 1 above)
Comment (nicolaszawada) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: Agreed, the first implementation was more verbose

File: lib/msal-browser/docs/caching.md (hunk 3 above)
Comment (nicolaszawada) [reply lib/msal-browser/docs/caching.md]: Agreed. I documented this more explicitly now.

File: change/@azure-msal-browser-25232dc8-c6d0-4238-ad59-1c1f3fcd1a7a.json (hunk 4 above)
Comment (nicolaszawada) [reply change/@azure-msal-browser-25232dc8-c6d0-4238-ad59-1c1f3fcd1a7a.json]: Added

File: lib/msal-browser/src/config/Configuration.ts (hunk 2 above)
Comment (nicolaszawada) [reply lib/msal-browser/src/config/Configuration.ts]: Added

**Comment by nicolaszawada** [comment]
> Thanks for the PR! Left a couple suggestions, would be great if you could include a couple tests as well.

I implemented your suggestions and added some tests.

File: lib/msal-browser/docs/caching.md [hunk 5]
```diff
@@ -58,10 +58,17 @@ To faciliate efficient token acquisition while maintaining a good UX, MSAL cache
...
+> :bulb: Temporary cache entries will always be stored in session storage or in memory unless overridden by the user with the `temporaryCacheLocation` cache option. MSAL will fallback to memory storage if local/session storage is not available. Please read the warning below for more information.
 
 > :bulb: The authorization code is only stored in memory and will be discarded after redeeming it for tokens.
 
+## Warning :warning:
+Overriding `temporaryCacheLocation` should be done with caution. Specifically when choosing `localStorage`. Interaction in more than one tab/window will not be supported and you may receive `interaction_in_progress` errors unexpectedly. This is an escape hatch, not a fully supported feature.
+
+When using MSAL in a scenario where the user is redirected after successful authentication in a new window or tab, the OAuth 2.0 Authorization Code with PKCE flow will be interrupted. In this case, the original window or tab where the authentication state (code verifier and challenge) are stored, will be lost, and the authentication flow will fail.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]:
```suggestion
When using MSAL.js with the default configuration in a scenario where the user is redirected after successful authentication in a new window or tab, the OAuth 2.0 Authorization Code with PKCE flow will be interrupted. In this case, the original window or tab where the authentication state (code verifier and challenge) are stored, will be lost, and the authentication flow will fail.
```

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts [hunk 6]
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
...
+
+        it("setTemporaryCache", () => {
+            browserSessionStorage.setTemporaryCache("cacheKey", cacheVal, true);
+            browserLocalStorage.setTemporaryCache("cacheKey2", cacheVal, true);
+            expect(window.sessionStorage.getItem(msalCacheKey)).toBe(cacheVal);
+            expect(window.localStorage.getItem(msalCacheKey2)).toBe(cacheVal);
+        });
+
+        it(
+            "getTemporaryCache falls back to local storage if not found in session/memory storage",
```
Comment (tnorling) [inline lib/msal-browser/test/cache/BrowserCacheManager.spec.ts]:
```suggestion
            "getTemporaryCache returns value from localStorage",
```

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts [hunk 7]
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
...
+        it("clear()", async () => {
+            browserSessionStorage.setTemporaryCache("cacheKey", cacheVal, true);
+            browserLocalStorage.setTemporaryCache("cacheKey", cacheVal, true);
+            await browserSessionStorage.clear();
+            await browserLocalStorage.clear();
+            expect(browserSessionStorage.getKeys()).toHaveLength(0);
+            expect(browserLocalStorage.getKeys()).toHaveLength(0);
+        });
+
+        describe("Getters and Setters", () => {
```
Comment (tnorling) [inline lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:770]: Tests below this line don't appear relevant to the changes in this PR. Am I missing something?

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts [hunk 8]
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
...
+                const testTempItemKey = "test-temp-item-key";
+                const testTempItemValue = "test-temp-item-value";
+                window.localStorage.setItem(testTempItemKey, testTempItemValue);
+                cacheConfig.cacheLocation = BrowserCacheLocation.LocalStorage;
+                browserLocalStorage = new BrowserCacheManager(TEST_CONFIG.MSAL_CLIENT_ID, cacheConfig, browserCrypto, logger);
+                expect(browserLocalStorage.getTemporaryCache(testTempItemKey)).toBe(testTempItemValue);
+            }
+        )
+
+        it("setItem", () => {
```
Comment (tnorling) [inline lib/msal-browser/test/cache/BrowserCacheManager.spec.ts]: This test doesn't appear relevant to this PR (probably already exists)

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts [hunk 9]
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
...
+                    browserSessionStorage.setRedirectRequestContext(testVal);
+                    expect(browserSessionStorage.getRedirectRequestContext()).toEqual(testVal);
+                });
+
+            });
+        });
+    });
+
+    
+    describe("Interface functions with overridden temporaryCacheLocation", () => {
```
Comment (tnorling) [inline lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:695]: Test additions above this line don't appear relevant to this PR

**Review by tnorling — COMMENTED** [review]
Thanks for addressing the feedback, code and docs look good. I'm a little confused about some of the tests, hoping you can clear that up. Thanks!

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 9 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:640]: I have copy-pasted these tests from "Interface functions" tests. But in this test setup I changed the cacheConfig for the temporaryCacheLocation to LocalStorage. The reasoning behind this was to show that the existing functionality doesn't break with the temporaryCacheLocation to LocalStorage.

Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:640]: If you prefer having them removed, let me know

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 8 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts]: Agreed

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 9 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:640]: These tests were already there

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 7 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:770]: I have copy-pasted these tests from the existing "Interface functions" tests. But in this test setup I changed the cacheConfig for the temporaryCacheLocation to LocalStorage. The reasoning behind this was to show that the existing functionality doesn't break with the temporaryCacheLocation to LocalStorage. If you prefer having them removed. Let me know

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 9 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:695]: These tests were already there. I just moved them in this file in a more logical order

File: lib/msal-browser/test/cache/BrowserCacheManager.spec.ts (hunk 6 above)
Comment (nicolaszawada) [reply lib/msal-browser/test/cache/BrowserCacheManager.spec.ts]: Agreed

File: lib/msal-browser/docs/caching.md (hunk 5 above)
Comment (nicolaszawada) [reply lib/msal-browser/docs/caching.md]: Agreed

**Comment by nicolaszawada** [comment]
> Thanks for addressing the feedback, code and docs look good. I'm a little confused about some of the tests, hoping you can clear that up. Thanks!

Thanks for your time & feedback, I have implemented your suggestions and cleared some things up. Let me know if there's anything else I can do.

**Comment by nicolaszawada** [comment]
@tnorling, I hope you're doing well. I understand that you may be busy with other tasks, but I just wanted to remind you that the PR is still open and awaiting review. If there's anything I can do to help, please don't hesitate to let me know.
Thank you for your time and consideration. I look forward to hearing back from you soon.

**Review by tnorling — APPROVED** [review]
Thanks for the reminder and addressing my feedback!

**Comment by nicolaszawada** [comment]
Thanks for the approval!
Hm I notice the msal-browser CI failing due to a timeout. I don't really see the issue. Is there anything I can do?

**Review by sameerag — APPROVED** [review]

**Comment by tnorling** [comment]
> 

Don't worry about it, tests have been a bit flakey lately, we'll sort it out

**Comment by nicolaszawada** [comment]
> > 
> 
> Don't worry about it, tests have been a bit flakey lately, we'll sort it out

Hi @tnorling, is there any update on the above please?

**Comment by nicolaszawada** [comment]
Hi @tnorling, I saw that there were some tests failing due to the fact that they were added **after** this PR and didn't include the new `temporaryCacheLocation` property. I have fixed the failing tests.

I noticed too late that you changed the target branch from `dev` to `msal-lts`, so I had to do some reverting cleanup of some `dev` commits that would have been merged into `msal-lts`. I think it would be better to squash commit this PR to not clutter the history.

**Comment by tnorling** [comment]
@nicolaszawada Thanks for fixing the tests and my apologies for the delay in merging
